### DISCUSSION TITLE
Does rendering contexts in root layout makes my entire application client side?

### DISCUSSION THREAD

**Discussion by arammaliachi** [vercel/next.js#58513] [App Router]
I am planning to start using SSR pages, but coming from Pages router i used to set a bunch of ContextProviders in my _app.js

From docs 
> With the provider rendered at the root, all other Client Components throughout your app will be able to consume this context.

If I render my ContextProviders in my root layout does that makes the rest of my application a client side app? If so what is the purpose of SSR if it can't support Context? Can't understand how this is meant to work

**Comment by icyJoseph** [comment]
No it does not.

I've started to use Josh Comeau's idea of owners, parents and children. Let's say you have:

```tsx
import ThemeProvider from './theme-provider'
 
export default function RootLayout({
  children,
}: {
  children: React.ReactNode
}) {
  return (
    <html>
      <body>
        <ThemeProvider>
           <Navigation />
           {children}
         </ThemeProvider>
      </body>
    </html>
  )
}
```

Where, the ThemeProvider is:

```tsx
'use client'
 
import { createContext } from 'react'
 
export const ThemeContext = createContext({})
 
export default function ThemeProvider({ children }) {
  return <ThemeContext.Provider value="dark">{children}</ThemeContext.Provider>
}
```

This means that, The RootLayout owns the ThemeProvider, and the Navigation, and it is also parent to the `children`, which in this case are the entire route contents.

The ThemeProvider can wrap a React sub tree, through the `children` prop. ThemeProvider is a parent to `children` but it doesn't own them, it is an isolated Client module (`use client`).

Since the RootLayout owns the ThemeProvider, the React Server, can serialize this VDOM representation, and put a slot/reference, where the ThemeProvider goes, pointing to some JavaScript bundle, and then it continues with the Navigation and the children, which for the root layout is the current Page contents, without opting in the entire subtree into client components. 

However, it is a different story within the ThemeProvider, since that's a client side module, through `use client`, now everything that it owns, will be a client component. Note that ThemeProvider doesn't own the `children` prop it gets passed, it is only the parent to that children prop.

And that's why doing this, make the entire application into a Client Component. Rather it helps to create a slot/reference in the serialized VDOM, to a client component, which injects a theme context provider. 

### One more thing

> the rest of my application a client side app?

I think here, you have been mislead, into thinking the `use client` means that Client Components are rendered, only on the client, but that's wrong.

Client Components, do get SSR'd, https://github.com/reactwg/server-components/discussions/4 ~ 

It is only Server Components that have the special behavior, which is to never render on the client, they do not generate a JavaScript bundle.
